Using BERTopic for topic analysis

sofiapinto opened the pull request:
---

# Description

In this PR we do topic analysis using BERTopic:
- we evaluate different models on different datasets (slices of MSE and BH forums);
- we then fix a random seed and check the results for specific models and interpret them;

Closes #3 

# Instructions for Reviewer
## Setup

In order to test the code in this PR you need to:
- clone this repo: `git clone git@github.com:nestauk/asf_public_discourse_home_decarbonisation.git`
- checkout to the correct branch: `git checkout 03_mse_bertopic`
- Run `make install`;
- Run `direnv allow`;
- Activate the conda enviroment: `conda activate asf_public_discourse_web_scraping`;
- Download spacy model: `python -m spacy download en_core_web_sm`

## Review

Hey @helloaidank, thank you for taking the time to review this PR.

**Scripts to be reviewed:**
There are a couple of scripts to be reviewed in this PR.
- Getters - I had to make changes to the getters, so please check if you think they make sense (this is also a result for changing the way BH data is extracted and stored - in another PR):
     - `asf_public_discourse_home_decarbonisation/getters/bh_getters.py`
     - `asf_public_discourse_home_decarbonisation/getters/mse_getters.py`
- Topic Analysis Pipeline (where the main files to be reviewed live)
    - `asf_public_discourse_home_decarbonisation/pipeline/bert_topic_analysis/bert_params_config.py` - config w/ parameters to test - I'm still not 100% happy about I organise the params to test (I don't like it that "text_column" and "filter" are in `model_and_additional_params`, but there are pros & cons about moving these to `data_source_parms`.
    - `asf_public_discourse_home_decarbonisation/pipeline/bert_topic_analysis/evaluate_bertopic_results.py` - script to evaluate BERTopic and compare results from different runs
- Topic analysis utils:
    - I created a few utils in  `asf_public_discourse_home_decarbonisation/utils/topic_analysis_utils.py`
- Topic Analysis - Analysis scripts: 
    - Feel free to open the scripts in `asf_public_discourse_home_decarbonisation/analysis/topic_analysis` as notebooks how I ran the topic analysis on specific slices of the data after evaluating different results.

**Code to run**
Feel free to test `asf_public_discourse_home_decarbonisation/pipeline/bert_topic_analysis/evaluate_bertopic_results.py`, but perhaps it's best if you comment some of the parms in `asf_public_discourse_home_decarbonisation/pipeline/bert_topic_analysis/bert_params_config.py` so that you can test it quickly.

# Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [ ] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [ ] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

helloaidank commented on asf_public_discourse_home_decarbonisation/pipeline/bert_topic_analysis/bert_params_config.py:
```diff
@@ -0,0 +1,110 @@
+from sklearn.feature_extraction.text import CountVectorizer
+
+# Additional parameters like the text column and filter are in the following dictionary instead of in the
+# data_source_parms because a different text column/filter would change the model specifities
```
I think this is the fine for the moment and we can try to think of another way of doing it later! I guess you could have an additional dictionaries for 'text_column' and 'filter' to avoid repetitiveness when we start looking at the text and original_posts?

sofiapinto replied:
I thought about that, but that also poses other problems, hence why I decided to keep it like this. Let's do as you say and keep it as it is for now

helloaidank commented on asf_public_discourse_home_decarbonisation/pipeline/bert_topic_analysis/evaluate_bertopic_results.py:
```diff
@@ -0,0 +1,367 @@
+- CONFIG_FILE_PATH is location of your configuration file in the repository structure
+e.g "asf_public_discourse_home_decarbonisation.pipeline.bert_topic_analysis.bert_params_config"
+Note you don't need the file extension .py at the end.
+- PATH_DATA if not reading standard forum data (e.g. if reading questions data)
```
nice!

helloaidank commented on asf_public_discourse_home_decarbonisation/pipeline/bert_topic_analysis/evaluate_bertopic_results.py:
```diff
@@ -0,0 +1,367 @@
+    """
+    config_module = importlib.import_module(path_to_config_file)
+
+    data_source_parms = config_module.data_source_parms
```
parms stands for parameters?

sofiapinto replied:
upsi... it was supposed to be params, short for parameters. not parms!

helloaidank commented on asf_public_discourse_home_decarbonisation/pipeline/bert_topic_analysis/evaluate_bertopic_results.py:
```diff
@@ -0,0 +1,367 @@
+    Creates and saves a figure with boxplots showing the distribution of results for:
+    - Number of topics
+    - Percentage of outliers
+    - Average probability of beloging to a non-outlier topic
```
```suggestion
    - Average probability of belonging to a non-outlier topic
```

helloaidank commented on asf_public_discourse_home_decarbonisation/pipeline/bert_topic_analysis/evaluate_bertopic_results.py:
```diff
@@ -0,0 +1,367 @@
+    percentage_of_outliers_df.boxplot(ax=axes[1], grid=False, vert=False)
+    axes[1].set_xlabel("Percentage of outliers")
+    # Making the x-axis always between 0 and 50 (just as visual aid)
+    if percentage_of_outliers_df.max().max() <= 50:
```
What happens if percentage of outliers is above 50%?

helloaidank commented on asf_public_discourse_home_decarbonisation/pipeline/bert_topic_analysis/evaluate_bertopic_results.py:
```diff
@@ -0,0 +1,367 @@
+    percentage_of_outliers_df.boxplot(ax=axes[1], grid=False, vert=False)
+    axes[1].set_xlabel("Percentage of outliers")
+    # Making the x-axis always between 0 and 50 (just as visual aid)
+    if percentage_of_outliers_df.max().max() <= 50:
```
If percentage of outliers is greater than 50%, does it plot with 0-100% axis I assume?

sofiapinto replied:
no, i think it plots from 0 to max found (or max found + a bit more)

helloaidank commented on asf_public_discourse_home_decarbonisation/pipeline/bert_topic_analysis/evaluate_bertopic_results.py:
```diff
@@ -0,0 +1,367 @@
+    avg_probablity_df.boxplot(ax=axes[2], grid=False, vert=False)
+    axes[2].set_xlabel("Average probablity of belonging to a non-outlier topic")
+
+    # axes[3].hist([res["outlier_fraction"] for res in outliers_vs_docs_dict.values()],
```
Delete this code or is it commented out for some reason?

sofiapinto replied:
forgot to leave a comment about this, sorry about that. So, remember that I wanted to also have a plot about the distirbution of outliers across runs? this is it, but it wasn't working well so I commented it out.

helloaidank commented on asf_public_discourse_home_decarbonisation/pipeline/bert_topic_analysis/evaluate_bertopic_results.py:
```diff
@@ -0,0 +1,367 @@
+
+def run_topic_model_evaluation(n_runs: int, docs: list, model_configs: dict) -> tuple:
+    """
+    Evaluates topic model on a specific number of runs.
```
Thinking potentially more of a description of what this function does would be nice. Such as what topic models are being run over?

helloaidank replied:
It could also be nice to modularise this function, it's quite big!

sofiapinto replied:
Sure, I'll look into doing both.

helloaidank commented on asf_public_discourse_home_decarbonisation/pipeline/bert_topic_analysis/evaluate_bertopic_results.py:
```diff
@@ -0,0 +1,367 @@
+    Returns:
+        pd.DataFrame: filtered data
+    """
+    if path_to_data_file is not None:
```
Is the "path_to_data_file" for some other form of data (that's not "mse" or "bh")?

sofiapinto replied:
yes, as written in the script docstring at the top of the file, it is used "if not reading standard forum data (e.g. if reading questions data)".
So if we want to run this on questions data, we need to pass the path to the questions data file. Let me know if you want me to add more context to this.

Also, we can change it to make it easier to read in questions' data. This was just a simple implementation I thought of, so that you could try and apply this to the questions data if you wanted to.

helloaidank replied:
Very well, sounds like a good idea.

helloaidank commented on asf_public_discourse_home_decarbonisation/pipeline/bert_topic_analysis/evaluate_bertopic_results.py:
```diff
@@ -0,0 +1,367 @@
+
+    # filter the data based on the keywords
+    if keywords is not None:
+        data = data[
```
Just for my understanding, then this will keep the data if a keyword is found for either the title column or the text column?

sofiapinto replied:
Yes - this is for the cases where we want to filter data by a set of keywords. If the keyword is found either in the title or the text column, then we keep that instance. This is interesting if you think about it, even if someone doesn't mention HPs in the title or the main post, but someone mentions HPs in a reply, we will still capture it.

helloaidank commented on asf_public_discourse_home_decarbonisation/pipeline/bert_topic_analysis/evaluate_bertopic_results.py:
```diff
@@ -0,0 +1,367 @@
+    for data_source in data_source_parms:
+        source_name = data_source["data_source"]
+
+        # One figure will be created for each slide of data considered
```
```suggestion
        # One figure will be created for each slice of data considered
```

helloaidank commented on asf_public_discourse_home_decarbonisation/analysis/topic_analysis/topic_analysis_heat_pump_keywords.py:
```diff
@@ -0,0 +1,227 @@
+from bertopic import BERTopic
+from umap import UMAP
+from bertopic.representation import KeyBERTInspired
+from sklearn.feature_extraction.text import CountVectorizer
```
KeyBERTInspired and CountVectorizer not required to run this script.
Same goes for pandas and matplotlib.pyplot

helloaidank commented on asf_public_discourse_home_decarbonisation/analysis/topic_analysis/topic_analysis_heat_pump_keywords.py:
```diff
@@ -0,0 +1,227 @@
+    .astype(str)
+    .apply(
+        lambda x: x.lower()
+        .replace("ashps", "ashp")
```
for ashps to ashp and gshps to gshp, why wouldn't we just replace ashps and gshps with "air source heat pumps" and "ground source heat pumps" respectively?

sofiapinto replied:
just made it this way so we always have the singular form for all of these acronyms, but no big difference in the end output.

helloaidank replied:
It won't make a big difference in the end to the output, only difference will be with sentences like "Why are there so many ashps" will go to "why are there so many air source heat pump" instead of "why are there so many air source heat pumps".

sofiapinto replied:
shouldn't be problematic for clustering titles, but i changed it now to:

```
mse_data["title"] = mse_data["title"].apply(
    lambda x: x.lower()
    .replace("ashps", "air source heat pumps")
    .replace("ashp", "air source heat pump")
    .replace("gshps", "ground source heat pumps")
    .replace("gshp", "ground source heat pump")
    .replace("hps", "heat pumps")
    .replace("hp", "heat pump")
    .replace("ufh", "under floor heating")
)
```

helloaidank commented on asf_public_discourse_home_decarbonisation/analysis/topic_analysis/topic_analysis_heat_pump_keywords.py:
```diff
@@ -0,0 +1,227 @@
+    width=1200,
+    height=750,
+    topics=range(15),
+    title="Top 15 topics in Heat Pump mentions in Builhub",
```
```suggestion
    title="Top 15 topics in Heat Pump mentions in Buildhub",
```

helloaidank commented on asf_public_discourse_home_decarbonisation/analysis/topic_analysis/topic_analysis_heat_pump_keywords.py:
```diff
@@ -0,0 +1,227 @@
+create_bar_plot_most_common_topics(topics_info=topics_info, top_n_topics=16)
+
+# %%
+topic_model.visualize_topics()
```
For all these, it would be great to be able to save the figures in a png so we can have a look at the outputs
e.g. 
`  fig = topic_model.visualize_topics()
    fig.write_image(figure_file_path + "topic_visualisation.png")`. I know this was initially done as a notebook, so I guess I can convert it back to that to view the images.

sofiapinto replied:
yep, at this point this is not the final script, so I didn't want to spend time writing code to save the figures. At the top of the file, in the docstring, I explain how to open these as notebooks in order to visualise the results.

helloaidank commented on asf_public_discourse_home_decarbonisation/analysis/topic_analysis/topic_analysis_mse_2021_vs_2022_vs_2023.py:
```diff
@@ -0,0 +1,224 @@
+# %%
+from bertopic import BERTopic
+from umap import UMAP
+from sklearn.feature_extraction.text import CountVectorizer
```
CountVectorizer, pandas and matplotlib.pyplot are not used

helloaidank commented on asf_public_discourse_home_decarbonisation/analysis/topic_analysis/topic_analysis_mse_categories.py:
```diff
@@ -0,0 +1,128 @@
+# %%
+from bertopic import BERTopic
+from umap import UMAP
+from sklearn.feature_extraction.text import CountVectorizer
```
CountVectorizer, pandas and matplotlib.pyplot not used

helloaidank commented:
Hi Sofia,

Thanks for doing this work, it's really great to have a visualisation of the evaluation and to double check that nothing too weird is going on with such a stochastic method.
Regarding the analysis scripts, it would be really nice to have them save their outputs to a relevant directory just so I can sense check them for myself. 

Let me know what you think!

sofiapinto commented:
Thanks @helloaidank. I'll leave the analysis scripts as they are for now, as they are still work in progress. I'll create a final script in a new PR where everything is optimised (i.e. you can use any keywords you want, instead of just HPs, etc) and save all figures to a nice location.

helloaidank commented:
> Thanks @helloaidank. I'll leave the analysis scripts as they are for now, as they are still work in progress. I'll create a final script in a new PR where everything is optimised (i.e. you can use any keywords you want, instead of just HPs, etc) and save all figures to a nice location.

Sounds good! Happy for you to merge :).

sofiapinto commented:
hey again @helloaidank ,

In the meantime:
- i've fixed that bug that only happened once in a while. Basically, when no topics were found we had 1 less output in our list, but the dataframe where the list was being stored expected the list to have a certain length and sometimes it didn't - hence why the code broke. So instead of dfs, I used dictionaries which are flexible in terms of length. That fixed it;
- Also implemented your suggestions;
- I removed the commented code about the outliers, I'll add it to a new PR.
